Add flex layout to Buttons block and new layout type.

tellthemachines opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Partially addresses #34872.

The Buttons block is slightly different to other blocks that have been migrated to layouts, because it has a Vertical variation that behaves in a flex column-like manner, and can be justified (though calling it justification is slightly inaccurate, because behind the scenes we're actually using `align-items` instead of `justify-content` to make it work with the column layout). 

This means that neither of the current layout options will work here: flow doesn't have flex properties, and flex is optimised for horizontal layouts. So I introduced a new "column" layout variation that uses `flex-direction: column` and works with justification controls, translating them to use `align-items`. 

I'd love to get some feedback on this approach before I start dealing with deprecations, tests etc. 😄 

I left the second part of #34872 (adding flex layout to the Navigation Block) out for now, because though it has a similar Vertical variation, the way styles are applied to the content of the block is much more complex, as we have several nested containers that need to be configured with the orientation and justification/alignment flex properties. Layout currently works by applying styles generically to the top level of the block, so we'll have to think about how to make that work with nested containers. Or if we want to make it work like that at all 😅 

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Add a buttons block to your editor. Check that horizontal and vertical orientation, plus all the relevant justification options, still work as expected.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

tellthemachines committed:
Add flex layout to Buttons block and new layout type.

oandregal commented:
Hey, I'm not sure if I can provide much help here. Riad, Joen, and Andrew Serong have been involved in a lot of this work (probably others too), so they'd probably have some thoughts hore.

jasmussen commented:
Thanks for the PR, took it for a spin. Just for comparison, here's a "before" GIF showing horizontal and vertical orientations with the justifications and spacing working:
![before](https://user-images.githubusercontent.com/1204802/138643283-8838216d-8afc-4b82-ab66-832ce874292c.gif)

Here's an "after" gif, showing the same but from this branch:

![buttons flex](https://user-images.githubusercontent.com/1204802/138643304-918e0190-4c67-4ad5-9484-9122528a13a4.gif)

Note one difference, here the justification controls are shown both in the toolbar and the inspector, which is the custom for blocks that opt into the proper flex method. That's good.

So just as a user experience and behavior, this appears to be working as intended 👍 👍

One thing I discovered. which could be due to my local environment, but sharing in case not:

- In trunk, create a vertical list of buttons
- Check out this branch, reload
- Note how the buttons are now horizontal

Here's the markup I tried, and it shows up as horizontal in this branch:

```
<!-- wp:buttons {"contentJustification":"right","orientation":"vertical","style":{"spacing":{"blockGap":"12px"}}} -->
<div class="wp-block-buttons is-content-justification-right is-vertical"><!-- wp:button /-->

<!-- wp:button /-->

<!-- wp:button /-->

<!-- wp:button /-->

<!-- wp:button /--></div>
<!-- /wp:buttons -->
```

andrewserong commented:
Nice idea @tellthemachines! The general approach sounds good to me, I like that the resulting CSS for the Buttons block is essentially unchanged 👍. There were a bunch of edge cases that @stacimc had to wrangle in #34546, so this looks like it shouldn't affect those changes, which is good.

> One thing I discovered. which could be due to my local environment, but sharing in case not:
> * In trunk, create a vertical list of buttons
> * Check out this branch, reload
> * Note how the buttons are now horizontal

I _think_ that sounds like it's because of the removal of the `orientation` attribute? From the PR description, it looks like @tellthemachines is planning to update the deprecations to handle that migration for existing blocks.

youknowriad commented on packages/block-editor/src/layouts/column.js:
```diff
@@ -0,0 +1,152 @@
+};
+
+export default {
+	name: 'column',
```
Do you think this layout will have other use-cases aside the "buttons" one? 
What happens when we apply this layout type to a group, does it bring any value? Do we want a variation like "Row" added for the group block with this layout?

ntsekouras replied:
I think of this `column` layout as a subtype of `flex` and its usefulness lies on the `justification` (at least for now).

The `default/flow` layout is another way to have a vertical list of blocks which now handles the `widths` of the block and targets `grouping blocks`.

Could this `justification` option be integrated with the `default` layout? Would it make sense? That would mean changing the `display` of `Group` (and other blocks that use the default layout) to `flex` and I'm not sure if there would be implications with that.

youknowriad replied:
> I think of this column layout as a subtype of flex and its usefulness lies on the justification (at least for now).

So potentially could also be implemented by introducing an "orientation" config to the "flex" layout itself.

ntsekouras replied:
That's another option for sure and that was my first thoughts about that when working with `flex` layout.

tellthemachines replied:
Thanks for the feedback! I've updated to use an "orientation" property on the flex layout. It makes sense because there was a lot of duplicated logic between flex and column.

tellthemachines committed:
Make orientation a property of flex layout

tellthemachines committed:
Update save function and add deprecation

tellthemachines committed:
Update fixtures

tellthemachines commented:
Thanks for the feedback, everyone! I've addressed comments, added a deprecation and updated the fixtures so this should be ready for a more in-depth review now 🙂

tellthemachines committed:
Fix php lint warnings.

youknowriad commented:
Took this for a spin. I noticed that when switching between buttons variations, I'm losing the "Justify config". It feels like we should probably keep it. In other words, an `orientation` config on the "flex" layout would only make sense if the other configs (at least most of them) are kept right? otherwise, maybe a separate layout is better. (Just trying to understand when do we add config to a layout VS when to create a separate layout and define guidelines around it)

It seems there's only "space between" that is not present on the vertical variation, while right now it's true that it doesn't make sense but for me it would still make sense there if we say introduce a "height" control later for the buttons block.

tellthemachines commented:
> Took this for a spin. I noticed that when switching between buttons variations, I'm losing the "Justify config". It feels like we should probably keep it.

Yeah, I looked into this and it's happening because of how we're setting the orientation as a block variation. The block variation defines e.g. `layout: { type: 'flex' }`, and it overrides the previous layout attribute. I'm thinking it probably makes more sense to set orientation as another Layout option in the block inspector, instead of as a block variation. I'll try that out and see how it works.

> It seems there's only "space between" that is not present on the vertical variation, while right now it's true that it doesn't make sense but for me it would still make sense there if we say introduce a "height" control later for the buttons block.

If we do add a height control, then perhaps we could make "space-between" display conditionally on the height being set? Because otherwise it won't have any effect, and the experience will feel broken.

tellthemachines committed:
Fix bad conflict resolution

tellthemachines committed:
Set orientation in layout tools

tellthemachines commented:
I've updated this to set orientation in the Layout section of the block inspector, instead of as a block variation. I also made it so that only blocks that have "orientation" defined in their layout attribute show the orientation controls, because they might not make sense for some blocks (I'm thinking of the Row variation of Group that should specifically be horizontal, but perhaps with this change we could potentially stop using Row altogether?)

jasmussen commented:
Lovely work as usual. I see this segmented control in the inspector now:

![buttons](https://user-images.githubusercontent.com/1204802/139810893-18529144-9041-4c82-ba3c-20434502ad9b.gif)

I'll defer to others including Riad on whether an orientation toggle is the correct one, but I do recall @shaunandrews already suggesting one, which is one vote in favor. He's also brought up that we should mostly avoid segmented controls with only two options. 

I think we can do better than the following sketch — pretty sure Shaun actually has a mockup somewhere we can leverage. But in the name of landing the _feature_ in the near term, such a toggle group with two arrows might be a good initial start that we can then polish further:

<img width="335" alt="Screenshot 2021-11-02 at 09 28 30" src="https://user-images.githubusercontent.com/1204802/139812242-7c028df5-0984-44d5-b812-e3871ac91a24.png">

We might even be able to show the two controls on a single line:

<img width="335" alt="One line" src="https://user-images.githubusercontent.com/1204802/139812298-acb81038-9b9d-4db0-b5e5-b983868cd3d2.png">

youknowriad commented on packages/block-editor/src/layouts/flex.js:
```diff
@@ -39,6 +52,12 @@ export default {
 					layout={ layout }
 					onChange={ onChange }
 				/>
+				{ layout?.orientation && (
```
This check is interesting. It basically means that if the current layout doesn't have any orientation, do not display the control to switch "orientation". That said one can argue that the "no orientation" is just equivalent to "horizontal orientation" meaning we could just always show the toggle.

The problem with always showing the toggle is that it will show up in other blocks using the "flex" layout like the "Group" block (or its Row variation and potentially others). While one could argue that this is wrong since is a "row" right, conceptually it's not wrong. For instance, what about blocks like "Social Icons" and "Query Pagination" also using the "flex" layout, should we allow them to switch orientation? It might make sense.

So the thing here is that I'm hesitant between:

 - Keep things as they are in the PR: making the "orientation" undefined or not in the default layout value the actual flag that indicates whether to show this control or not.
 - Introduce a dedicated opt-out flag in the layout support config (and potentially only use it in the Group block)

tellthemachines replied:
An opt-out flag would be nice, especially if there's only one block where we really don't want to enable vertical orientation. Perhaps something like `"allowOrientation": false`?

It also occurred to me that if orientation were enabled for all blocks, we wouldn't need Row as a variation of Group at all. But Group is using flow layout; would there be any issues in migrating it to flex with vertical orientation?

ZebulanStanphill replied:
Well, a flexbox layout doesn't let children use `float`, so yes, there would definitely be issues if "flow" did not remain as an option.

tellthemachines replied:
Oooh good point. I guess we'd better leave it as it is then.

youknowriad reviewed (approved):
Regardless of my questioning of the opt-out behavior, this is looking very good.

tellthemachines committed:
Improve orientation controls

tellthemachines committed:
Change orientation to opt-out

tellthemachines commented:
I updated the inspector controls and tried switching orientation to opt out; it seems to work well! Any further comments, feedback etc.?

ntsekouras commented on packages/block-library/src/buttons/edit.js:
```diff
@@ -58,31 +33,16 @@ function ButtonsEdit( {
-					} }
-				/>
-			</BlockControls>
+			<BlockControls
```
We can remove this, no?

ntsekouras commented on packages/block-library/src/buttons/deprecated.js:
```diff
@@ -1,9 +1,79 @@
+		return attributes;
+	}
+
+	const { contentJustification, orientation } = attributes;
```
We should remove these attributes here and keep only the rest. You can also see them appear in the fixtures.

tellthemachines replied:
I'm not sure I understand. These are used here to migrate any legacy settings to the new layout, e.g. if there's a deprecated block with "orientation" set we use that value in layout when we update the block. Or is there another way of doing it?

In the fixtures these attributes only appear in the deprecated versions of Buttons. That should be expected, right?

tellthemachines replied:
Oh wait, you mean I should be removing them from the updated attributes?

ntsekouras replied:
> I should be removing them from the updated attributes?

Yes. We extract them here and use them for the `layout` but since the current version doesn't contain them, these attributes shouldn't exist at the final migrated version of the block.

ntsekouras commented on packages/block-library/src/buttons/deprecated.js:
```diff
@@ -1,9 +1,79 @@
+				},
+			},
+		},
+		isEligible: ( { layout } ) => ! layout,
```
I don't think we should have `isEligible` here. The thing is that when  this function is checked it gets the `attributes` for this deprecation. This ends up having a newly created `buttons` block (using its default layout) and then in `migrate` the `orientation:horizontal` exists and applies the defaults from the migration.

You can check this by adding a different `default` in `buttons` block.json like:
```
"default": {
	"type": "flex",
	"orientation": "vertical",
	"justifyContent": "right"
}
```

Add a `Buttons` block, save and then reload.

tellthemachines replied:
Oooh, interesting. How does it work in the [Social Links deprecation](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/social-links/deprecated.js#L86)?

I can't remove `isEligible` altogether or the deprecation won't work, but I'll change the check to explicitly look for the `contentJustification` and `orientation` attributes, as these are the only ones we need to migrate. If they don't exist we can leave the block as is.

tellthemachines replied:
All feedback has now been addressed in #36192. Thanks for the review!

ntsekouras replied:
>How does it work in the Social Links deprecation?

Because:
>The specific handling by `className` below is needed because `itemsJustification`
was introduced in https://github.com/WordPress/gutenberg/pull/28980/files and wasn't
declared in block.json.

So I didn't remove any attributes there.

ntsekouras reviewed (commented):
Thanks so much for your great work here @tellthemachines! 

Sorry for being late catching up, but I've left some comments to discuss 😄

shaunandrews commented:
Ha, @jasmussen here's the mockup I had from a few months back:

<img width="339" alt="image" src="https://user-images.githubusercontent.com/191598/140101137-45092268-d8b8-4020-b5d0-7ea4c3588e30.png">

The biggest difference from what you posted about is the language. I find "Direction" and "Alignment" to be more easily understandable than "Orientation" and "Justification," which feel more like jargon.

--

What I see in `trunk` is this:

<img width="266" alt="image" src="https://user-images.githubusercontent.com/191598/140102374-4673ebdc-257d-4e44-8436-186b7c9f8b95.png">

I feel like the Orientation should be shown first, with the Justification settings after — and that these two controls are connected; If I change the Orientation to vertical, the Justification icons should rotate and their descriptions should update to use top and bottom instead of left and right. Similarly, I'm not sure if the "multiple lines" toggle makes much sense when the Orientation is set to vertical.

Why doesn't the vertical orientation support space-between? I guess it would require the container to have a defined height?

Here's how I would expect it to look:

<img width="618" alt="image" src="https://user-images.githubusercontent.com/191598/140103342-478934db-b887-46f8-91b6-85589ed6def9.png">

tellthemachines commented:
One question: can we now remove the [legacy alignment styles](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/buttons/style.scss) for Buttons or should they remain for back-compat purposes?

jasmussen commented:
> One question: can we now remove the legacy alignment styles for Buttons or should they remain for back-compat purposes?

I think we've tried a few times (@kjellr is so busy these days he might not have time to respond here), and we've always come back to not being able to. We might be in the future, but if you're able to keep them for the time being, it's probably for the best.

kjellr commented:
Yeah, we've tried to remove those in the past, but they've broken existing content. I'd avoid it here if we can.

TJBriggs commented:
Is it still possible to define contentJustification in our block templates? Did the syntax change? When I try to center a button within my code like this:

`['core/buttons', { className: 'hero__buttons', contentJustification: 'center' }, [
	['core/button', { className: 'hero__button', text: 'Button Text' }],
]],`

The blocks load initially but after publishing and refreshing the page in the admin, the buttons block breaks and I get the following error in the console:

`Error: Minified React error #185; visit https://reactjs.org/docs/error-decoder.html?invariant=185 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.`

`Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.`

tellthemachines commented:
The syntax for content justification did change; instead of `{contentJustification: 'center' }` it is now `{ layout: {type: 'flex', justifyContent: 'center' } }`. That said, there shouldn't be any errors with using the old syntax because there's a deprecation in place to handle it. 

Could you share some reproduction steps and the version of Gutenberg you're on?

youknowriad commented:
@tellthemachines unfortunately deprecations don't work for templates, this has always been the case in Gutenberg. It's not really clear what the solution forward for this is. It might require some structural or API changes ultimately.
